src/mpt.rs

```diff
@@ -0,0 +1,660 @@
+/*
+    Copyright © 2023, ParallelChain Lab
+    Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+//! This mod only public to crate inside. Provides structs and implementations
+
+use crate::db::{KeyInstrumentedDB, DB};
+use crate::error::MptError;
+use crate::trie_key::PrefixedTrieNodeKey;
+use crate::version::{Version, VersionProvider};
+use hash_db::{AsHashDB, HashDB, HashDBRef, Hasher as KeyHasher, Prefix};
+use pchain_types::cryptography::Sha256Hash;
+use reference_trie::{ExtensionLayout, NoExtensionLayout, RefHasher};
+use std::collections::{HashMap, HashSet};
+use trie_db::proof::generate_proof;
+use trie_db::{Trie, TrieDBBuilder, TrieDBMutBuilder, TrieMut};
+
+type Hash256 = [u8; 32];
+
+pub type Proof = Vec<Vec<u8>>;
+const NULL_NODE_KEY: &[u8] = &[0_u8];
+const NULL_NODE_VALUE: &[u8] = &[0_u8];
+
+/// `MptChanges` is a wrapper of changes in [Mpt] when call function close()
+#[derive(Debug, Clone)]
+pub(crate) struct MptChanges(
+    pub(crate) HashMap<Vec<u8>, Vec<u8>>,
+    pub(crate) HashSet<Vec<u8>>,
+    pub(crate) Sha256Hash,
+);
+
+/// `Mpt` is struct to maintain the Merkle Patricia Trie tree as storage struct
+///
+/// Merkle Tree: A hash tree in which each node’s hash is computed from its child nodes hashes.
+///
+/// Patricia Trie: A efficient Radix Trie (r=16), a data structure in which “keys” represent the path one has to take to reach a node
+#[derive(Debug, Clone)]
+pub(crate) struct Mpt<'a, S, V>
+where
+    S: DB + Send + Sync + Clone,
+    V: VersionProvider + Send + Sync + Clone,
+{
+    storage: KeyInstrumentedDB<'a, S, V>,
```

Comment:
I think it'd be better if we use `db` terminology to refer to the physical key-value store that pchain-world-state stores its data instead of `storage` terminology, since "storage" can also refer to storage tries.